vtparse/src/transitions.rs

```diff
@@ -0,0 +1,343 @@
+// Builds up the transition table for a state machine based on
+// https://vt100.net/emu/dec_ansi_parser
+
+use crate::enums::{Action, State};
+
+/// Returns 1 if a <= x and x <= b. Otherwise returns 0.
+const fn in_range(x: u8, a: u8, b: u8) -> u8 {
+    ((x <= b) as u8) & ((a <= x) as u8)
+}
+
+/// Returns `true_value` if `condition` is true. Otherwise returns `false_value`.
+const fn cond(condition: bool, true_value: u8, false_value: u8) -> u8 {
+    ((condition as u8) * true_value) | ((!condition as u8) * false_value)
+}
+
+/// Match `u8` using given patterns. Return `OptionPack`.
+/// The patterns should not overlap.
+macro_rules! match_action_state {
+    ( $name:ident => {
+        $( $a:tt $( ..= $b:tt )? => ($action:ident, $state:ident), )*
+    }) => {
+        OptionPack(
+            $(
+                ({
+                    const B: u8 = [$a $(,$b)?][[$a $(,$b)?].len() - 1];
```

Comment:
This is pretty hard work to understand at a glance; could you add some comments to explain it?
I *think* I can follow it, but it's been a solid 15 minutes of reading and thinking :)

Reply:
(this actually looks like perl code!)